base/sort.jl

```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+            t[j] = x                  # put the element where it belongs
+            counts[i] = j + 1         # increment the target index for the next
+        end                           #  ↳ element in this bucket
+
+        v, t = t, v # swap the now sorted destination list t back into the primary list v
+
+    end
+
+    v
+end
+
+function sort!(v::AbstractVector{<:Bool}, lo::Integer, hi::Integer, a::AdaptiveSort, o::Ordering)
+    first = lt(o, false, true) ? false : lt(o, true, false) ? true : return v
+    count = 0
+    for i in lo:hi
+        if v[i] == first
+            count += 1
+        end
+    end
+    v[lo:lo+count-1] .= first
+    v[lo+count:hi] .= !first
+    v
+end
+
+maybe_unsigned(x::Integer) = x # this is necessary to avoid calling unsigned on BigInt
+maybe_unsigned(x::Union{Int8, Int16, Int32, Int64, Int128}) = unsigned(x)
+function _extrema(v::AbstractArray, lo::Integer, hi::Integer, o::Ordering)
+    mn = mx = v[lo]
+    while lo < hi
+        lo += 1
+        vi = v[lo]
+        lt(o, vi, mn) && (mn = vi)
+        lt(o, mx, vi) && (mx = vi)
+    end
+    mn, mx
+end
+function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::AdaptiveSort, o::Ordering)
+    # if the sorting task is unserializable, then we can't radix sort or sort_int_range!
+    # so we skip straight to the fallback algorithm which is comparison based.
+    U = Serial.Serializable(eltype(v), o)
+    U === nothing && return sort!(v, lo, hi, a.fallback, o)
+
+    # to avoid introducing excessive detection costs for the trivial sorting problem
+    # and to avoid overflow, we check for small inputs before any other runtime checks
+    hi <= lo && return v
+    lenm1 = maybe_unsigned(hi-lo) # adding 1 would risk overflow
+    # only count sort on a short range can compete with insertion sort fo lenm1 < 30
+    # and the optimization is not worth the detection cost, so we use insertion sort.
+    lenm1 < 30 && return sort!(v, lo, hi, SMALL_ALGORITHM, o)
+
+    # UInt128 does not support fast bit shifting so we never
+    # dispatch to radix sort but we may still perform count sort
+    if sizeof(U) > 8
+        if eltype(v) <: Integer && o isa DirectOrdering
+            v_min, v_max = _extrema(v, lo, hi, Forward)
+            v_range = maybe_unsigned(v_max-v_min)
+
+            # we know lenm1 ≥ 30, so this will never underflow.
+            # if lenm1 > 3.7e18 (59 exabytes), then this may incorrectly dispatch to fallback
+            if v_range < 5lenm1-100 # count sort will outperform comparison sort if v's range is small
+                return sort_int_range!(v, Int(v_range+1), v_min, o === Forward ? identity : reverse, lo, hi)
+            end
+        end
+        return sort!(v, lo, hi, a.fallback, o)
+    end
+
+    v_min, v_max = _extrema(v, lo, hi, o)
+    if eltype(v) <: Integer && o isa DirectOrdering
+        R = o === Reverse
+        v_range = maybe_unsigned(R ? v_min-v_max : v_max-v_min)
```

Comment:
Because it is possible that the range/span overflows in the check for dispatch to count sort which would result in a critical dispatch error (time & space complexity > `typemax(Int)`). Once that dispatch branch is triggered, we know that the range/span is less than the length of the input vector. Assuming that the input vector has length less than `typemax(Int)`, it's safe to convert this to `Int`. If not, the conversion throws an error. I'm pretty sure this assumption is implicit everywhere, but I can't test it because I don't have a 32-bit machine that can handle a 4GB vector, nor a 64-bit machine that can handle an 18EB vector.

Concretely, without this unsigned, we would dispatch to count sort on `rand(Int, 300)` because `v_max-v_min < 0 < div(lenm1, 2)`

Reply:
Ah of course. FWIW, you can just run the 32-bit Julia binaries on a 64-bit machine to check that 4GB vectors are handled correctly.

Reply:
Vectors that large are not handled correctly.
`Vector{Int}(undef, 2^29)` or larger gives an invalid array size error.
`BitVector(undef, Int64(2)^32)` or larger gives an inexact error truncating to `Int32`.
Given these results, I think it's fair to assume the return type of `length` is an `Int` and throw an inexact error if it is larger than `typemax(Int)`.